### PR TITLE
[Merged by Bors] - refactor: extract and name the pregroupoid underlying contDiffGroupoid

### PR DESCRIPTION
**Pull request by grunweg**
This will be used for proving the inverse function theorem on manifolds. 

--------

To avoid duplication, I'll prove a conceptual version (stated in terms of `G.isLocalStructomorphWithinAt`) first and deduce versions for `C^n` and analytic maps later; their hypothesis are a suitable property of the pregroupoid inducing `G`.
Applying this to the smooth or analytic category entails extracting the underlying pregroupoid.

### CODE REVIEW

File: Mathlib/Geometry/Manifold/SmoothManifoldWithCorners.lean [hunk 1]
```diff
@@ -529,46 +529,51 @@ variable {m n : ℕ∞} {𝕜 : Type*} [NontriviallyNormedField 𝕜] {E : Type*
 
 variable (n)
 
-/-- Given a model with corners `(E, H)`, we define the groupoid of `C^n` transformations of `H` as
-the maps that are `C^n` when read in `E` through `I`. -/
+/-- Given a model with corners `(E, H)`, we define the pregroupoid of `C^n` transformations of `H`
+as the maps that are `C^n` when read in `E` through `I`. -/
+def contDiffPregroupoid : Pregroupoid H := {
+  property := fun f s => ContDiffOn 𝕜 n (I ∘ f ∘ I.symm) (I.symm ⁻¹' s ∩ range I)
+  comp := fun {f g u v} hf hg _ _ _ => by
```
Comment (ocfnash) [inline Mathlib/Geometry/Manifold/SmoothManifoldWithCorners.lean:536]:
```suggestion
def contDiffPregroupoid : Pregroupoid H where
  property f s := ContDiffOn 𝕜 n (I ∘ f ∘ I.symm) (I.symm ⁻¹' s ∩ range I)
  comp {f g u v} hf hg _ _ _ := by
```

File: Mathlib/Geometry/Manifold/SmoothManifoldWithCorners.lean [hunk 2]
```diff
@@ -529,46 +529,51 @@ variable {m n : ℕ∞} {𝕜 : Type*} [NontriviallyNormedField 𝕜] {E : Type*
...
+    · rintro x ⟨hx1, _⟩
+      simp only [mfld_simps] at hx1 ⊢
+      exact hx1.2
+  id_mem := by
+    apply ContDiffOn.congr contDiff_id.contDiffOn
+    rintro x ⟨_, hx2⟩
+    rcases mem_range.1 hx2 with ⟨y, hy⟩
+    rw [← hy]
+    simp only [mfld_simps]
+  locality := fun {f u} _ H => by
```
Comment (ocfnash) [inline Mathlib/Geometry/Manifold/SmoothManifoldWithCorners.lean:551]:
```suggestion
  locality {f u} _ H := by
```

File: Mathlib/Geometry/Manifold/SmoothManifoldWithCorners.lean [hunk 3]
```diff
@@ -529,46 +529,51 @@ variable {m n : ℕ∞} {𝕜 : Type*} [NontriviallyNormedField 𝕜] {E : Type*
...
+    rw [← hx] at hy1 ⊢
+    simp only [mfld_simps] at hy1 ⊢
+    rcases H x hy1 with ⟨v, v_open, xv, hv⟩
+    have : I.symm ⁻¹' (u ∩ v) ∩ range I = I.symm ⁻¹' u ∩ range I ∩ I.symm ⁻¹' v := by
+      rw [preimage_inter, inter_assoc, inter_assoc]
+      congr 1
+      rw [inter_comm]
+    rw [this] at hv
+    exact ⟨I.symm ⁻¹' v, v_open.preimage I.continuous_symm, by simpa, hv⟩
+  congr := fun {f g u} _ fg hf => by
```
Comment (ocfnash) [inline Mathlib/Geometry/Manifold/SmoothManifoldWithCorners.lean:564]:
```suggestion
  congr {f g u} _ fg hf := by
```

File: Mathlib/Geometry/Manifold/SmoothManifoldWithCorners.lean [hunk 4]
```diff
@@ -529,46 +529,51 @@ variable {m n : ℕ∞} {𝕜 : Type*} [NontriviallyNormedField 𝕜] {E : Type*
...
+    apply hf.congr
+    rintro y ⟨hy1, hy2⟩
+    rcases mem_range.1 hy2 with ⟨x, hx⟩
+    rw [← hx] at hy1 ⊢
+    simp only [mfld_simps] at hy1 ⊢
+    rw [fg _ hy1]
+}
```
Comment (ocfnash) [inline Mathlib/Geometry/Manifold/SmoothManifoldWithCorners.lean:571]:
```suggestion
```

**Review by ocfnash — APPROVED** [review]
Thanks!

Please apply the suggestions and then feel free to merge.

bors d+

**Comment by grunweg** [comment]
Thanks for the review! I merged master and pushed one last change (inlining a trivial subgoal).
bors r+

**Pull request closed without merging**